library/ecp.c

```diff
@@ -2249,7 +2258,7 @@ static unsigned char ecp_pick_window_size( const mbedtls_ecp_group *grp,
      * (The last test is useful only for very small curves in the test suite.)
      */
 #if( MBEDTLS_ECP_WINDOW_SIZE < 6 )
-    if( w > MBEDTLS_ECP_WINDOW_SIZE )
+    if( (!p_eq_g || !ecp_group_is_static_comb_table(grp)) && w > MBEDTLS_ECP_WINDOW_SIZE )
```

Comment:
The window size returned by this function determines the size of comb table.

When MBEDTLS_ECP_FIXED_POINT_OPTIM is enabled, we use static comb table.
But the size of the static comb table is fixed and generated based on the
default config (MBEDTLS_ECP_WINDOW_SIZE == 6).
If the user reduces MBEDTLS_ECP_WINDOW_SIZE in config, the pre-computed static
comb table still have its fixed size. So we should return the size of
pre-computed static comb table, and it should not affected by user config.

And when p_eq_g == 0, the comb table is always dynamically allocated and
calculated. So let the user choice the max size of comb table.

Some comments are added to describe this condition.

Reply:
I think the code is correct but I also agree it's not very easy to read. Unfortunately I'm not sure I can find a way to make it more readable. Trying to sum up, there are 3 cases:

1. If `!p_eq_g` with use a dynamically-created table of multiples of the input point, stored in a local variable (not in `grp->T`)
2. If `p_eq_g` there a two cases:
   a. If we have a static table in `grp->T`, we use it
   b. otherwise, we create it dynamically and store it in `grp->T` (hoping it will be re-used)

The only case where we want to allow `w` to be greater than `MBEDTLS_ECP_WINDOW_SIZE` is 2a (static table).

Perhaps the confusion comes from the fact that there can be two comb tables: a transient one, always dynamic, stored in a local variable, and a permanent one stored in `grp->T`, which this PR is making static in most cases. And the function `ecp_group_is_static_comb_table()` only tells us whether the permanent table is static or dynamic, but it doesn't tell use if that's the table we're going to be using for the current operation: for that we need to check if `p_eq_g`.